embedded/sql/engine.go

```diff
@@ -1113,100 +1215,143 @@ func (e *Engine) QueryPreparedStmt(stmt *SelectStmt, params map[string]interface
 		return nil, ErrIllegalArguments
 	}
 
-	e.catalogRWMux.RLock()
-	defer e.catalogRWMux.RUnlock()
+	e.mutex.RLock()
+	defer e.mutex.RUnlock()
+
+	if e.closed {
+		return nil, ErrAlreadyClosed
+	}
 
 	// TODO (jeroiraz): won't be needed when in-memory catalog becomes transactional
 	if e.catalog == nil {
 		return nil, ErrCatalogNotReady
 	}
 
 	if renewSnapshot {
-		err := e.RenewSnapshot()
+		err := e.renewSnapshot()
 		if err != nil && err != tbtree.ErrReadersNotClosed {
 			return nil, err
 		}
 	}
 
-	snapshot, err := e.Snapshot()
+	snapshot, err := e.getSnapshot()
 	if err != nil {
 		return nil, err
 	}
 
-	implicitDB, err := e.DatabaseInUse()
+	implicitDB, err := e.databaseInUse()
 	if err != nil {
 		return nil, err
 	}
 
-	_, _, _, err = stmt.compileUsing(e, implicitDB, params)
+	_, err = stmt.compileUsing(e, implicitDB, params)
 	if err != nil {
 		return nil, err
 	}
 
 	return stmt.Resolve(e, implicitDB, snapshot, params, nil)
 }
 
-func (e *Engine) ExecStmt(sql string, params map[string]interface{}, waitForIndexing bool) (ddTxs, dmTxs []*store.TxMetadata, err error) {
+func (e *Engine) ExecStmt(sql string, params map[string]interface{}, waitForIndexing bool) (summary *ExecSummary, err error) {
 	return e.Exec(strings.NewReader(sql), params, waitForIndexing)
 }
 
-func (e *Engine) Exec(sql io.ByteReader, params map[string]interface{}, waitForIndexing bool) (ddTxs, dmTxs []*store.TxMetadata, err error) {
+func (e *Engine) Exec(sql io.ByteReader, params map[string]interface{}, waitForIndexing bool) (summary *ExecSummary, err error) {
 	stmts, err := Parse(sql)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
 	return e.ExecPreparedStmts(stmts, params, waitForIndexing)
 }
 
-func (e *Engine) ExecPreparedStmts(stmts []SQLStmt, params map[string]interface{}, waitForIndexing bool) (ddTxs, dmTxs []*store.TxMetadata, err error) {
-	e.catalogRWMux.Lock()
-	defer e.catalogRWMux.Unlock()
+type ExecSummary struct {
+	DDTxs []*store.TxMetadata
+	DMTxs []*store.TxMetadata
+
+	UpdatedRows     int
```

Comment:
In this case is not matched but counting the number of updated/inserted ones. Even if the values assigned to the columns are the same, it will produce a new version of the row with a different internal timestamp.